test/jdk/java/lang/String/CompactString/MaxSizeUTF16String.java

```diff
@@ -118,4 +118,33 @@ public void testMaxCharArray() {
             }
         }
     }
+
+    /*
+     * Test that UTF-8 of too large strings throws OOME, (not NegativeArraySizeException).
+     */
+    @Test
+    public void testMaxUTF8_UTF16Encode() {
+        String s = "\uFFFF";
+        final byte[] bytes1 = s.getBytes(StandardCharsets.UTF_8);
+        assertEquals(3, bytes1.length, "UTF_8 encoded length of 0xffff");
+
+        int min = Integer.MAX_VALUE / bytes1.length - 1;
+        int max = min + 3;
+
+        // String of size min can be UTF_8 encoded.
+        System.out.println("testing size: " + min);
+        String s1 = s.repeat(min);
+        byte[] bytes = s1.getBytes(StandardCharsets.UTF_8);
+        int remaining = Integer.MAX_VALUE - bytes.length;
+        assertTrue(remaining >= bytes1.length, "remainder too large: " + remaining);
+
+        // Strings of size min+1...min+2, throw OOME
+        // The resulting byte array would exceed implementation limits
+        for (int count = min + 1; count < max; count++) {
```

Comment:
The case `min + 1` cannot lead to a `NegativeArraySizeException` in the current code, since `3 * (min + 1) <= MAX_VALUE`. In theory, it should succeed by returning the encoded `byte[]`, although It throws `OOME` for exceeding VM limits. That is, this case does not trigger the invocation of `computeSizeUTF8_UTF16()` in the proposed fix.

Only `min + 2` throws `NegativeArraySizeException` in the current code, and thus the invocation of `computeSizeUTF8_UTF16()` in the proposed fix.

Reply:
Indeed, different OOMEs are thrown in the two cases triggered by different limits, min +2 is due to integer overflow, while min +1  is due a VM limit on the size of byte[Integer.MAX_VALUE]. Different VM implementations may have different limits. on the max size of a byte array.

Reply:
There might be some merit in lowering the threshold at which an exact size computation is triggered.
The oversized allocation "wastes" quite a bit of memory and causes extra GC work and usually triggers a second copy of the final size.
However, some guess or heuristic would be needed to choose the threshold at which extra cpu work is needed to compute the exact size vs some metric as to the "cost" of wasted memory (and saving on the copy).
Most guesses would be somewhat arbitrary; bigger than 1Mb, 1GB, etc....? 
Choosing that number would be out of scope for the issue raised by this bug.

Reply:
What I mean is that the VM limit might change and suddenly accept `MAX_VALUE` as an allowed array size (very unlikely, I guess). The test would then fail on `min + 1` because it expects a OOME which will not be thrown.

But that is very remote.